ipn/ipnlocal/serve.go

```diff
@@ -271,154 +276,67 @@ func (b *LocalBackend) DeleteForegroundSession(sessionID string) error {
 		return nil
 	}
 	sc := b.serveConfig.AsStruct()
+	if hp, ok := shouldDeleteFunnel(sc, sessionID); ok {
+		delete(sc.AllowFunnel, hp)
+		if len(sc.AllowFunnel) == 0 {
+			sc.AllowFunnel = nil
+		}
+	}
 	delete(sc.Foreground, sessionID)
 	return b.setServeConfigLocked(sc)
 }
 
-// StreamServe opens a stream to write any incoming connections made
-// to the given HostPort out to the listening io.Writer.
-//
-// If Serve and Funnel were not already enabled for the HostPort in the ServeConfig,
-// the backend enables it for the duration of the context's lifespan and
-// then turns it back off once the context is closed. If either are already enabled,
-// then they remain that way but logs are still streamed
-func (b *LocalBackend) StreamServe(ctx context.Context, w io.Writer, req ipn.ServeStreamRequest) (err error) {
-	f, ok := w.(http.Flusher)
-	if !ok {
-		return errors.New("writer not a flusher")
-	}
-	f.Flush()
-
-	port, err := req.HostPort.Port()
-	if err != nil {
-		return err
-	}
-
-	// Turn on Funnel for the given HostPort.
-	sc := b.ServeConfig().AsStruct()
-	if sc == nil {
-		sc = &ipn.ServeConfig{}
+// shouldDeleteFunnel returns the port and true if:
+// 1. This foreground has a TCP port, that
+// 2. Funnel has it referenced and allowed, and
+// 3. No other foreground or background session has the port
+// Ambiguity: what if another background had funnel on?
+func shouldDeleteFunnel(sc *ipn.ServeConfig, sessionID string) (hp ipn.HostPort, ok bool) {
```

Comment:
this func might be a good candidate for a table of tests. It's a pure function of its inputs and it's kinda non-trivial.

Reply:
Per our chat, we removed this function since we're nested AllowFunnel in the foreground configs.